components/script/dom/document.rs

```diff
@@ -1465,6 +1466,34 @@ impl Document {
         self.window.reflow(ReflowGoal::Full, ReflowReason::KeyEvent);
     }
 
+    pub fn dispatch_composition_event(
+        &self,
+        composition_event: ::keyboard_types::CompositionEvent,
+    ) {
+        let focused = self.get_focused_element();
+        let body = self.GetBody();
+
+        let target = match (&focused, &body) {
+            (&Some(ref focused), _) => focused.upcast(),
+            (&None, &Some(ref body)) => body.upcast(),
+            (&None, &None) => self.window.upcast(),
+        };
```

Comment:
Where is the spec for that?

Reply:
I mirrored the code from keyboard events but it turns out that the [spec says](https://w3c.github.io/uievents/#event-type-compositionstart):

> Event.target : focused element processing the composition 

Am I right that composition events should only be emitted if some element is focused and otherwise discarded?

Reply:
Changed it to conform to spec.

Reply:
Please add a link for this bit of the spec here.

Reply:
Done.